File: hrms-frontend/src/services/timesheetService.js
// src/services/timesheetService.js
import api from '../api/axiosConfig';

const timesheetService = {
  getAll: () => api.get('/timesheets'),
  getById: (id) => api.get(`/timesheets/${id}`),
  getByEmployee: (empId, params) =>
    api.get(`/timesheets/employee/${empId}`, { params }),
  getByProject: (projectId) => api.get(`/timesheets/project/${projectId}`),
  getByStatus: (status) => api.get(`/timesheets/status/${status}`),
  create: (data) => api.post('/timesheets', data),
  update: (id, data) => api.put(`/timesheets/${id}`, data),
  delete: (id) => api.delete(`/timesheets/${id}`),

  submit: (id) => api.patch(`/timesheets/${id}/submit`),
  approve: (id) => api.patch(`/timesheets/${id}/approve`),
  reject: (id, reason) =>
    api.patch(`/timesheets/${id}/reject`, { reason }),
  submitBulk: (ids) => api.post('/timesheets/bulk/submit', ids),
  approveBulk: (ids) => api.post('/timesheets/bulk/approve', ids),

  getEmployeeStats: (empId) => api.get(`/timesheets/stats/employee/${empId}`),
  getProjectStats: (projectId) =>
    api.get(`/timesheets/stats/project/${projectId}`),
  getWeekly: (empId, weekStart) =>
    api.get(`/timesheets/weekly/${empId}?weekStart=${weekStart}`),
};

export default timesheetService;

File: hrms-backend/src/main/java/com/hrms/service/TimesheetService.java

//service/TimesheetService.java
package com.hrms.service;

import com.hrms.dto.response.TimesheetResponse;
import com.hrms.model.Timesheet;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface TimesheetService {
 TimesheetResponse create(Timesheet timesheet);
 TimesheetResponse update(Long id, Timesheet timesheet);
 TimesheetResponse getById(Long id);
 List<TimesheetResponse> getAll();
 List<TimesheetResponse> getByEmployee(Long employeeId);
 List<TimesheetResponse> getByEmployeeAndDateRange(
     Long employeeId, LocalDate start, LocalDate end);
 List<TimesheetResponse> getByProject(Long projectId);
 List<TimesheetResponse> getByStatus(String status);
 void delete(Long id);

 // Workflow
 TimesheetResponse submit(Long id);
 TimesheetResponse approve(Long id, Long approverId);
 TimesheetResponse reject(Long id, String reason);
 void submitBulk(List<Long> ids);
 void approveBulk(List<Long> ids, Long approverId);

 // Analytics
 BigDecimal getTotalHours(Long employeeId, LocalDate start, LocalDate end);
 Map<String, Object> getEmployeeStats(Long employeeId);
 Map<String, Object> getProjectStats(Long projectId);
 Map<String, Object> getWeeklySummary(Long employeeId, LocalDate weekStart);
}